### PR TITLE
[Dependency Scanner] Refactor the global scanning service to no longer maintain scanner cache state

### PR DESCRIPTION
**Pull request by artemcm**
Instead, each scan's 'ModuleDependenciesCache' will hold all of the data corresponding to discovered module dependencies.

The initial design presumed the possibility of sharing a global scanning cache amongst different scanner invocations, possibly even different concurrent scanner invocations.

This change also deprecates two libSwiftScan entry-points: `swiftscan_scanner_cache_load` and `swiftscan_scanner_cache_serialize`. They never ended up getting used, and since this code has been largely stale, we are confident they have not otherwise had users, and they do not fit with this design.

A follow-up change will re-introduce module dependency cache serialization on a per-query basis and bring the binary format up-to-date.

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

File: test/ScanDependencies/module_deps_cache_reuse.swift [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // RUN: %empty-directory(%t)
 // RUN: mkdir -p %t/clang-module-cache
+// XFAIL: *
```
Comment (cachemeifyoucan) [inline test/ScanDependencies/module_deps_cache_reuse.swift:3]: Can you add comments for all XFAIL?

**Review by cachemeifyoucan — APPROVED** [review]

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test macOS platform

**Comment by artemcm** [comment]
@swift-ci test linux platform

**Comment by artemcm** [comment]
@swift-ci test linux platform

**Pull request merged**